picard/acoustid/manager.py

```diff
@@ -39,55 +39,83 @@ def __init__(self, fingerprint, duration, orig_recordingid=None, recordingid=Non
+        return int((len(self.fingerprint) + len(self.puid)) * 1.1)
+
+    def is_submitted(self):
+        return not self.recordingid or self.orig_recordingid == self.recordingid
 
 
 class AcoustIDManager(QtCore.QObject):
 
-    # AcoustID has a post limit of around 1 MB. With the data submitted by
-    # Picard this is roughly around 250 fingerprints. Submit a few less to have
-    # some leeway.
-    BATCH_SUBMIT_COUNT = 240
+    # AcoustID has a post limit of around 1 MB.
+    MAX_PAYLOAD = 1000000
+    # Limit each submission to N attempts
+    MAX_ATTEMPTS = 5
+    # In case of Payload Too Large error, batch size is reduced by this factor
+    # and a retry occurs
+    BATCH_SIZE_REDUCTION_FACTOR = 0.7
 
     def __init__(self, acoustid_api):
         super().__init__()
-        self._fingerprints = {}
+        self._submissions = {}
         self._acoustid_api = acoustid_api
 
     def add(self, file, recordingid):
         if not file.acoustid_fingerprint or not file.acoustid_length:
             return
         puid = file.metadata['musicip_puid']
-        self._fingerprints[file] = Submission(file.acoustid_fingerprint, file.acoustid_length, recordingid, recordingid, puid)
+        self._submissions[file] = Submission(file.acoustid_fingerprint, file.acoustid_length, recordingid, recordingid, puid)
         self._check_unsubmitted()
 
     def update(self, file, recordingid):
-        submission = self._fingerprints.get(file)
+        submission = self._submissions.get(file)
         if submission is None:
             return
         submission.recordingid = recordingid
         self._check_unsubmitted()
 
     def remove(self, file):
-        if file in self._fingerprints:
-            del self._fingerprints[file]
+        if file in self._submissions:
+            del self._submissions[file]
         self._check_unsubmitted()
 
     def is_submitted(self, file):
-        submission = self._fingerprints.get(file)
+        submission = self._submissions.get(file)
         if submission:
-            return not submission.recordingid or submission.orig_recordingid == submission.recordingid
+            return submission.is_submitted()
         return True
 
     def _unsubmitted(self):
-        for file, submission in self._fingerprints.items():
-            if submission.recordingid and submission.orig_recordingid != submission.recordingid:
+        for file, submission in self._submissions.items():
+            if not submission.is_submitted():
                 yield (file, submission)
 
+    def _batch(self, submissions):
+        size = 0
+        for file, submission in submissions:
+            if submission.attempts > self.MAX_ATTEMPTS:
+                continue
+            size += len(submission)
+            if size > self.max_batch_size:
+                log.debug("AcoustID: Max batch size attained: %d > %d", size, self.max_batch_size)
+                break
+            submission.attempts += 1
```

Comment:
Nice detail here is that retry is per fingerprint submission. But submission.attempts should get reset I think if one hits the submit action again.